### PR TITLE
Allow Pokemon names as arguments in /stone

### PR DESCRIPTION
**Pull request by PartMan7**
Suggested in https://www.smogon.com/forums/threads/stone-pokemon-name-instead-of-stone-stone-name.3657629/#post-8620547

Behaviour:

Displays the stone for Pokemon with a single Mega Evolution (Beedrill, Bee, Beedrill-Mega)
Displays all stones if there are multiple Mega Evolutions (Mewtwo, Mewtwo-Mega)
Displays the relevant stone if a particular Mega Evolution is specified (Mewtwo-Mega-X)
Displays Dragon Ascent for Rayquaza-Mega (Rayquaza, Ray, Rayquaza-Mega)

All warnings have been retained.

### CODE REVIEW

**Comment by PartMan7** [comment]
Oh, and I've added a message for Orbs in Gen 8+ instead of just this.parse(`/help stone`).

File: server/chat-plugins/othermetas.ts [hunk 1]
```diff
@@ -187,91 +187,116 @@ export const commands: ChatCommands = {
...
+			if (!megas.length) return this.errorReply(`Error: Mega Evolution not found.`);
+			stones = megas.map(poke => {
+				const megaPoke = dex.getSpecies(poke);
+				const flag = megaPoke.requiredMove === 'Dragon Ascent' ? megaPoke.requiredMove : megaPoke.requiredItem;
+				if (/mega[xy]$/.test(targetid) && toID(megaPoke.name) !== toID(dex.getSpecies(targetid))) return null;
+				if (!flag) {
+					console.log(megaPoke);
+					console.log(flag);
+					return null;
+				}
```
Comment (AnnikaCodes) [inline server/chat-plugins/othermetas.ts:208]:
```suggestion
				if (!flag) return null;
```

Comment (PartMan7) [reply server/chat-plugins/othermetas.ts:208]: Oh, right.

File: server/chat-plugins/othermetas.ts [hunk 2]
```diff
@@ -187,91 +187,116 @@ export const commands: ChatCommands = {
...
+				return getMegaStone(flag, sep[1]);
+			}).filter(poke => poke !== null);
+			if (!stones.length) return this.errorReply(`Error: Mega Evolution not found.`);
+		}
+		const toDisplay = (stones || [stone]), banlist = Dex.getFormat('gen8mixandmega').banlist;
```
Comment (AnnikaCodes) [inline server/chat-plugins/othermetas.ts:209]:
```suggestion
		const toDisplay = (stones || [stone]);
		const banlist = Dex.getFormat('gen8mixandmega').banlist;
```

File: server/chat-plugins/othermetas.ts [hunk 3]
```diff
@@ -187,91 +187,116 @@ export const commands: ChatCommands = {
...
+		toDisplay.forEach(aStone => {
```
Comment (AnnikaCodes) [inline server/chat-plugins/othermetas.ts:211]: As per [the contributing guidelines](https://github.com/smogon/pokemon-showdown/blob/master/CONTRIBUTING.md#es5-and-es6), this should be a `for-of` loop instead.

File: server/chat-plugins/othermetas.ts [hunk 4]
```diff
@@ -187,91 +187,113 @@ export const commands: ChatCommands = {
...
+			if (!species.otherFormes) return this.errorReply(`Error: Mega Evolution not found.`);
+			const megas = species.otherFormes.filter(poke => /(?:-Primal|-Mega(?:-[XY])?)$/.test(poke));
+			if (!megas.length) return this.errorReply(`Error: Mega Evolution not found.`);
+			stones = megas.map(poke => {
+				const megaPoke = dex.getSpecies(poke);
+				const flag = megaPoke.requiredMove === 'Dragon Ascent' ? megaPoke.requiredMove : megaPoke.requiredItem;
+				if (/mega[xy]$/.test(targetid) && toID(megaPoke.name) !== toID(dex.getSpecies(targetid))) return null;
+				if (!flag) return null;
+				return getMegaStone(flag, sep[1]);
+			}).filter(poke => poke !== null);
```
Comment (AnnikaCodes) [inline server/chat-plugins/othermetas.ts:206]: I don't think that looping over it twice (in `.map` and `.filter`) is the best plan — a for loop would be just as readable IMO.

File: server/chat-plugins/othermetas.ts [hunk 5]
```diff
@@ -187,91 +187,113 @@ export const commands: ChatCommands = {
...
+			if (!stones.length) return this.errorReply(`Error: Mega Evolution not found.`);
+		}
+		const toDisplay = (stones || [stone]);
 		const banlist = Dex.getFormat('gen8mixandmega').banlist;
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:209]:
```suggestion
		const ruleTable = Dex.getRuleTable(Dex.getFormat('gen8mixandmega'));
```

File: server/chat-plugins/othermetas.ts [hunk 6]
```diff
@@ -190,23 +190,22 @@ export const commands: ChatCommands = {
 		if (stone && dex.gen >= 8 && ['redorb', 'blueorb'].includes(stone.id)) {
 			return this.errorReply("The Orbs do not exist in Gen 8 and later.");
 		}
-		let stones;
+		let stones = [];
```
Comment (AnnikaCodes) [inline server/chat-plugins/othermetas.ts:193]:
```suggestion
		const stones = [];
```

File: server/chat-plugins/othermetas.ts [hunk 7]
```diff
@@ -190,23 +190,22 @@ export const commands: ChatCommands = {
...
 		if (!stone) {
 			const species = dex.getSpecies(targetid.replace(/(?:mega[xy]?|primal)$/, ''));
 			if (!species.exists) return this.errorReply(`Error: Mega Stone not found.`);
 			if (!species.otherFormes) return this.errorReply(`Error: Mega Evolution not found.`);
-			const megas = species.otherFormes.filter(poke => /(?:-Primal|-Mega(?:-[XY])?)$/.test(poke));
-			if (!megas.length) return this.errorReply(`Error: Mega Evolution not found.`);
-			stones = megas.map(poke => {
+			for (let poke of species.otherFormes) {
```
Comment (AnnikaCodes) [inline server/chat-plugins/othermetas.ts:198]:
```suggestion
			for (const poke of species.otherFormes) {
```

File: server/chat-plugins/othermetas.ts [hunk 8]
```diff
@@ -187,91 +187,112 @@ export const commands: ChatCommands = {
...
-		} else {
-			buf += `<span class="col itemiconcol"><psicon item="${targetid}"/></span> `;
-		}
-		if (targetid === "dragonascent") {
-			buf += `<span class="col movenamecol" style="white-space:nowrap"><a href="https://${Config.routes.dex}/moves/${targetid}" target="_blank">Dragon Ascent</a></span> `;
-		} else {
-			buf += `<span class="col pokemonnamecol" style="white-space:nowrap"><a href="https://${Config.routes.dex}/items/${stone.id}" target="_blank">${stone.name}</a></span> `;
+		for (const aStone of toDisplay) {
+			if (!aStone) return;
+			if (banlist.includes(aStone.name)) {
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:212]:
```suggestion
			if (ruleTable.isBanned(`${aStone.name === 'Dragon Ascent' ? 'move' : 'item'}:${aStone.name}`)) {
```

Comment (KrisXV) [reply server/chat-plugins/othermetas.ts:212]: Though honestly, we don't really need the warning messages anymore

**Comment by Zarel** [comment]
Uh, @TheImmortal I guess?

**Review by KrisXV — APPROVED** [review]

**Pull request merged**